Beta distribution

guyko81 opened the pull request:
This is a non-working version of Beta-Bernoulli distribution. At first iteration it optimises the trees but stops right after. Probably caused by the singular matrix that stops np.linalg.solve. np.linalg.lstsq doesn't throw error but can't solve the optimisation further.

Pull request only for easier review, not for actual merge.

I added an example notebook with Kaggle Give Me Some Credit competition for testing purposes.

tonyduan commented:
Hi @guyko81,

Thanks again for the PR and apologies for the delay in response. 

I haven't tested the derivations below so If you could attach your derivations for discussion, that'd be much appreciated.

**Beta**

![Screen Shot 2020-06-07 at 5 22 53 PM](https://user-images.githubusercontent.com/1163598/83983457-86278180-a8e3-11ea-82dd-e5d579585ad6.png)

In particular, I'd replace lines 16-17 of `beta.py` with the following (to address missing parens, typos, signs).
```
        D[:, 0] = -self.alpha * (digamma(self.alpha + self.beta) - digamma(self.alpha) + np.log(Y))
        D[:, 1] = -self.beta * (digamma(self.alpha + self.beta) - digamma(self.beta) + np.log(1 - Y))
```

**Beta-Bernoulli**

![Screen Shot 2020-06-07 at 5 36 08 PM](https://user-images.githubusercontent.com/1163598/83983687-6002e100-a8e5-11ea-9ac8-9ec4e326c697.png)

Let me know if this works for you (perhaps first starting with the Beta, then the Beta-Bernoulli). Since here you're fixing `n=1` in the Beta-Bernoulli model, we might want to rename it to something else, but we can discuss further after it starts working.

guyko81 commented:
I have updated my code with the derivatives above but still doesn't work. If I'm not mistaken the problem with the natural gradient is that the derivative is symmetrical and so the det(D) is zero, that's why the singular matrix every case, regardless of the data. 

A quick help for everyone doing derivatives (I found Wolfram Alpha quite hard to use). I think it would be useful to mention in the guide too.
from sympy.stats import BetaBinomial, density
from sympy import log
from sympy import *

a = Symbol('a')
b = Symbol('b')
x = Symbol('x')
dist = BetaBinomial('dist', n=1, alpha=exp(a), beta=exp(b))

# d/d(a)
factor(expand(-log(density(dist).pmf(x)).diff(a)))
# d/d(b)
factor(expand(-log(density(dist).pmf(x)).diff(b)))

Funny fact: without the negative sign the model was able to calculate a rough estimation of the outcome average (alpha is constant for all cases, beta is constant for all cases, but they are different from each other), but with the negative sign the outcome became a 50% prediction (alpha=beta, but at least with different values for all cases). 

Other thing is that I was able to build the same model in Tensorflow with neural network and it works there. So there's a good gradient flow for the problem, I just don't understand the error in NGBoost.
This was my loss in Tensorflow:
def beta_bernoulli_maximum_likelihood(y_true, y_pred):
    alpha, beta = tf.split(value=y_pred, num_or_size_splits=[1, 1], axis=1, name='coef_split') 
    betabernoulli_dist = tfp.distributions.BetaBinomial(total_count=tf.ones_like(alpha), concentration1=alpha, concentration0=beta)
    return K.mean(K.binary_crossentropy(y_true, betabernoulli_dist.mean()), axis=-1)

tonyduan commented:
@guyko81 Could you push your updated code to your branch? I'm unable to see any changes reflected in the codebase for `D_score` of either the Beta distribution or the BetaBernoulli.

> If I'm not mistaken the problem with the natural gradient is that the derivative is symmetrical and so the det(D) is zero, that's why the singular matrix every case, regardless of the data.

I'm not sure what you mean by "the derivative is symmetrical", can you elaborate? In the branch I can't find an implementation of the `metric` function for either the Beta or the BetaBernoulli logarithmic scores. Can you point me to where it's implemented? By default the code will use a Monte Carlo estimate of the Fisher Information, but for precision and performance we'd prefer to implement it explicitly. 

Finally, the `fastbetabinom` package fails to install on my OS X because it requires Cython and OpenMP. In the interest of keeping our package lean it'd be preferable to implement a custom `fit` function that performs maximum likelihood estimation in the BetaBernoulli distribution instead of calling `fastbetabinom`.

guyko81 commented:
@tonyduan I have updated the code, sorry for not pushing in my repository before.
When I say the derivative is symmetrical I mean the d/da(score) and d/db(score). I print out the metric and grad values
                print(np.linalg.cond(metric))
                print(metric[0])
                print(grad[0])
and it shows the problem: the metric has too high condition number.

I'm rewriting the fit function and pushing again.

guyko81 commented:
I just updated the fit function based on fastbetabino code. You should be able to run now.
Referring back to the det(D) problem, here is the print output for my run:

print(np.linalg.cond(metric))
[1.27381033e+16 1.27381033e+16 1.27381033e+16 ... 1.27381033e+16
 1.27381033e+16 1.27381033e+16]

print(metric[0])
[[ 0.25 -0.25]
 [-0.25  0.25]]

print(grad[0])
[ 0.5 -0.5]

So the problem (again,, only as per my understanding) is the metric. The matrix is symmetrical and det(D) is zero, so np.linalg.solve will drop error: 
LinAlgError: Singular matrix

tonyduan commented:
@guyko81 Looks to me like this is happening because the `metric` function isn't defined for the BetaBernoulli and the fallback empirical Fisher information isn't working. This is a limitation of using the empirical Fisher information which I'll have to look into once I find more time.

For now, your code runs on my machine once I set `natural_gradient=False` in NGBoost, which avoids using the empirical Fisher. Let me know how that works for you.

Moving forward I strongly recommend implementation of the analytical Fisher information in the `metric` function of `BetaBernoulliLogScore`.

tonyduan commented on ngboost/distns/betabernoulli.py:
```diff
@@ -0,0 +1,123 @@
+        # self.beta = self.m * (1 - self.mu) # m = alpha + beta
+        # self.dist = dist(n=1, a=self.alpha, b=self.beta)
+        self.log_alpha = params[0]
+        self.log_beta = params[0]
```
Also, there's a typo here this should be set to `params[1]`.

guyko81 replied:
Oh, that typo! Thanks! It works for me now too! 
I will try to make the Fisher information in the metric function and start a clean pull request when it works.
